### ISSUE TITLE
Pulumi is incompatible with Azure CLI 2.53.1

### ISSUE THREAD

**Issue by bb33f** [pulumi/pulumi#14431]
### What happened?

UPDATE/WORK-AROUND:  Downgrading Azure CLI to 2.49.0 has resolved the problem.

I installed Pulumi v3.91.0 today.  When I execute "pulumi preview" on a project/stack that has been working for months, I get the following error:

error: getting stack configuration: get stack secrets manager: open keeper azurekeyvault://REDACTED.vault.azure.net/keys/REDACTED: failed to Dial default KeyVault: Invoking Azure CLI failed with the following error: 'az' is not recognized as an internal or external command,
operable program or batch file.

### Example

see above

### Output of `pulumi about`

'dotnet build -nologo .' completed successfully
CLI
Version      3.91.0
Go Version   go1.21.3
Go Compiler  gc

Plugins
NAME               VERSION
azure-native       1.104.0
dotnet             unknown
pulumi.azuread     4.2.1
pulumi.kubernetes  3.5.2

Host
OS       Microsoft Windows 11 Pro
Version  10.0.22621 Build 22621
Arch     x86_64

This project is written in dotnet: executable='C:\Program Files\dotnet\dotnet.exe' version='7.0.402'

Backend
Name           REDACTED
URL            azblob://REDACTED
User           REDACTED
Organizations
Token type     personal

Dependencies:
NAME                VERSION
pulumi.azuread      4.2.1
Pulumi.AzureNative  1.104.0
pulumi.kubernetes   3.5.2

Pulumi locates its logs in C:\Users\REDACTED
warning: Failed to get information about the current stack: constructing secrets manager of type "cloud": open keeper azurekeyvault://REDACTED.vault.azure.net/keys/REDACTED: failed to Dial default KeyVault: Invoking Azure CLI failed with the following error: 'az' is not recognized as an internal or external command,
operable program or batch file.

### Additional context

None of my ENV variables have changed and I have verified that AZURE_KEYVAULT_AUTH_VIA_CLI=true.
I confirmed that I am logged into az.
I verified that az commands work from several directories so my PATH is okay.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by shahiddev** [comment]
I can confirm I'm seeing same issue - appears to be specifically impacting PowerShell - a previously working environment now throws this error. 
I switched to WSL and it works fine.
AZ CLI version 2.53.1 

I can confirm `$env:AZURE_KEYVAULT_AUTH_VIA_CLI = "true"` is present (it's in the PowerShell profile)

**Comment by bb33f** [comment]
Same issue with pulumi 3.91.1

**Comment by Megasware128** [comment]
After some testing the problem doesn't seem to be caused by a Pulumi update so it might be caused by an azure-cli or PowerShell update. I get this same issue if I downgrade back to 3.70.0 and I'm pretty sure it worked once.

**Comment by shahiddev** [comment]
I downgraded to 2.53.0 and its working for me now

**Comment by justinvp** [comment]
It's not clear to me what changed in Azure CLI 2.53.1 that causes the problem. Here's the diff compared to 2.53.0: https://github.com/Azure/azure-cli/compare/azure-cli-2.53.0...azure-cli-2.53.1

**Comment by shahiddev** [comment]
> It's not clear to me what changed in Azure CLI 2.53.1 that causes the problem. Here's the diff compared to 2.53.0: [Azure/azure-cli@azure-cli-2.53.0...azure-cli-2.53.1](https://github.com/Azure/azure-cli/compare/azure-cli-2.53.0...azure-cli-2.53.1)

Agreed it's not obvious, my colleagues have also seen this issue. 
I just upgraded back to 2.53.1 and it works now 🤷 (is this the proverbial switching it off and on again).

If it matters I'm using Choco to manage the dependencies

**Comment by Megasware128** [comment]
Downgrading to

> I downgraded my AZ CLI to 2.53.0 and its working for me now with Pulumi 3.91.0 (EDIT: and 3.91.1)

I tried downgrading and it didn't work for me. Going as far back as 2.40.0. I'm using winget btw

**Comment by Megasware128** [comment]
Upon conducting further tests, I discovered that Pulumi was unable to locate 'az' on the path. However, placing 'az' in the working directory resolved this. Yet, this led to another issue where the azure-native provider would malfunction due to the presence of a relative 'az' executable. My current workaround involves adding 'az' to the working directory, and swiftly deleting it once Pulumi begins processing, but prior to the initialization of the azure-native provider. This workaround, albeit functional, lends itself to a rather inconvenient workflow 😅.

**Comment by alecor191** [comment]
I'm having the same issue as others in this thread. Unfortunately, downgrading Az CLI and/or Pulumi didn't help for me.

This is as far as I got with my investigation by profiling process activity:

- When running `pulumi` commands, the Pulumi process calls 
  ```bash
  C:\Windows\system32\cmd.exe /c az account get-access-token -o json --resource https://vault.azure.net
  ```
- The `cmd` process is launched with the following `PATH` environment variable:
  ```
  PATH=;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin; C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin
  ```
  Notice the leading `;` and the `<space>` between the two directories.
- If I set the path above and try to run the `cmd` command above, I get the same error. Only when I remove the leading `;` and the mentioned `<space>`, then the `cmd` command executes successfully.
- Therefore, (in my case) I believe the issue is a malformed `PATH` env var used when calling Az CLI.

The `pulumi` process launches with the `PATH` values that I have configured as my user + system environment variables. So it appears as if Pulumi would set `PATH` explicitly for the `cmd` process it launches.

I'm now trying to find out where these paths are coming from...if anyone has more context on this, I would love to get your thoughts.

**Comment by alecor191** [comment]
**TL;DR**: Workaround: Setting env var `AzureCLIPath` to `C:\Program Files\Microsoft SDKs\Azure\CLI2\wbin` solved the issue for me.

### Context
Pulumi appears to [use](https://github.com/pulumi/go-cloud/blob/master/secrets/azurekeyvault/akv.go#L197) the library [go-autorest](https://github.com/Azure/go-autorest).

[That library sets](https://github.com/Azure/go-autorest/blob/main/autorest/azure/cli/token.go#L218) the `PATH` environment variable by composing a list of paths from a few different parts that results in the malformed `PATH`, \____if___\_ some of those parts are empty.

That code hasn't been touched in 2 years though; nor did the Pulumi code using it. So why do we see the issue now?

This is the part I'm still trying to understand: Looking at the relevant `go-autorest` code:

```go
	// This is the path that a developer can set to tell this class what the install path for Azure CLI is.
	const azureCLIPath = "AzureCLIPath"

	// The default install paths are used to find Azure CLI. This is for security, so that any path in the calling program's Path environment is not used to execute Azure CLI.
	azureCLIDefaultPathWindows := fmt.Sprintf("%s\\Microsoft SDKs\\Azure\\CLI2\\wbin; %s\\Microsoft SDKs\\Azure\\CLI2\\wbin", os.Getenv("ProgramFiles(x86)"), os.Getenv("ProgramFiles"))

        ...
		cliCmd = exec.Command(fmt.Sprintf("%s\\system32\\cmd.exe", os.Getenv("windir")))
		cliCmd.Env = os.Environ()
		cliCmd.Env = append(cliCmd.Env, fmt.Sprintf("PATH=%s;%s", os.Getenv(azureCLIPath), azureCLIDefaultPathWindows))
```

`os.Environ()` would have to return an empty string to get to the mentioned faulty `PATH` string. My suspicion is that `os.Environ()` used to return the values of the `PATH` env var set in Windows. And recently something changed, that caused this call to return an empty string (even though I confirmed that the Pulumi process is launched with paths listed in the `PATH` env var).

**Comment by Megasware128** [comment]
@alecor191 Great find! Better then my workaround. Looks like the lib also has gone out of support.

**Comment by justinvp** [comment]
Thanks for the deep dive, @alecor191! We've run into several issues when attempting upgrading `gocloud.dev` in the recent past that required reverting the upgrade. We need to increase our test coverage before another attempt (https://github.com/pulumi/pulumi/issues/11986). This issue may be a good reason to prioritize that effort.

Glad that there's a workaround of setting `AzureCLIPath` in the meantime.
